[BE] feat: 홈화면API 추천 카테고리에 IT트렌드와 랜덤 카테고리 추가

yuyoungrhee opened the pull request:
## 📌 변경 내용 & 이유

기존 추천 카테고리 로직의 책임 분리, 성능 개선, 비즈니스 로직 변경을 위해 전체적으로 리팩터링을 진행했습니다.

### 1. 추천 카테고리 비즈니스 로직 변경
- **AS-IS**
    - 회원: 북마크 기반 추천 3개
    - 비회원: 하루 동안 고정된 랜덤 추천 3개
- **TO-BE**
    - **회원**: IT 트렌드(1) + 북마크 기반(3) + 랜덤(1) = **총 5개**
    - **비회원**: IT 트렌드(1) + 랜덤(4) = **총 5개**
    - 랜덤 카테고리는 이제 **API를 호출할 때마다 새로운 결과**를 반환합니다.

### 2. `RecommendCategoryService` 분리
- **이유**: 기존 `HearitService`에 있던 추천 로직이 복잡해짐에 따라, **카테고리를 추천하는 책임**을 새로운 `RecommendCategoryService`로 완전히 분리했습니다.
- **개선 후**: `HearitService`는 추천된 카테고리 목록으로 히어릿을 조회하는 역할에만 집중할 수 있게 되어, 코드의 가독성과 테스트 용이성이 향상되었습니다.

### 3. 랜덤 추천 후보 조회 성능 개선
- 기존에는 DB에서 모든 카테고리 ID를 조회한 후 애플리케이션 메모리에서 필터링했으나, `NOT IN` 절을 사용하는 쿼리(`findIdsWithoutExcludedIds`)를 추가하여 **DB 레벨에서 필터링**하도록 변경했습니다.
- 불필요한 데이터를 DB에서 애플리케이션으로 전송하지 않으므로, 데이터가 많아질수록 성능상 이점을 가집니다.

## 🧪 테스트 방법 (선택)
- `RecommendCategoryServiceTest`
- `HearitServiceTest`
- `HearitIntegrationTest`

## 📢 논의하고 싶은 내용
- 추천 로직을 별도 서비스로 분리한 현재 구조에 대해 다른 의견이 있으시면 편하게 공유해주세요!
- ex) 추천카테고리별 히어릿 조회 서비스 자체를 다른 서비스로 분리하자 등...

## 🧩 관련 이슈
close #626 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - 추천 카테고리 API 신설: GET /api/v1/recommendations/categories. 로그인 없이도 접근 가능(회원은 개인화 반영). 카테고리 개수(categorySize)와 항목 개수(hearitSize) 쿼리 파라미터 지원, 범위를 벗어나면 기본값으로 자동 보정. 카테고리별 추천 히어릿 목록을 반환.
- Refactor
  - 추천 기능을 전용 컨트롤러로 분리. 기존 경로(/api/v1/hearits/recommend-category)는 하위 호환을 위해 계속 응답하나, 신규 경로 사용을 권장.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yuyoungrhee committed:
[BE] feat: 추천 카테고리에 IT트렌드와 랜덤히어릿 추가 #626

yuyoungrhee committed:
[BE] refactor: 메서드 분리 및 네이밍 개선 #626

yuyoungrhee committed:
[BE] Merge branch 'develop-be' into feat/#626-add-it-trend-and-random-categories-to-recommended-hearit #626

# Conflicts:
#	backend/app/src/main/java/com/onair/hearit/app/hearit/application/HearitService.java

yuyoungrhee committed:
[BE] test: CategoryRepository 새로운 메서드 테스트 추가 #626

yuyoungrhee commented:
@coderabbitai 리뷰해줘

yuyoungrhee commented on backend/app/src/main/java/com/onair/hearit/app/category/application/RecommendCategoryService.java:
```diff
@@ -0,0 +1,86 @@
+            recommendations.addAll(getUserBasedRecommendations(userInfo, recommendations));
+        }
+        recommendations.addAll(getRandomRecommendations(recommendations));
+        return recommendations;
```
UserInfo가 로직상 null이될 수가 없는데 처리해야됩니까?

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/app/hearit/application/HearitServiceTest.java:
```diff
@@ -177,77 +183,38 @@ void getHearitDetailNotFoundTest() {
+                    .orElseThrow();
+            softly.assertThat(itTrendResponse.hearits()).extracting(HearitResponse::hearitId)
+                    .containsExactlyInAnyOrder(hearit_it1.getId(), hearit_it2.getId());
         });
```
HearitServiceTest에서 RecommendCategoryService의 구현까지 테스트하게 되는 나쁜 테스트 아닙니까?

yuyoungrhee committed:
[BE] refactor: 추천 기능 별도 도메인&패키지 분리 및 api 개선 #626

yuyoungrhee committed:
[BE] fix(auth): recommendations api 인증필터 화이트리스트에 추가 #626

yuyoungrhee commented on backend/app/src/main/java/com/onair/hearit/app/auth/config/ApiSecurityConfig.java:
```diff
@@ -34,7 +34,8 @@ public class ApiSecurityConfig {
             "/api/*/hearits/**",
             "/api/*/categories/**",
             "/api/*/keywords/**",
-            "/api/*/playing-histories/**"
+            "/api/*/playing-histories/**",
+            "/api/*/recommendations/**",
```
JwtAuthenticationFilter에서 아래와 같이 처리하고있어서 괜찮음

```java
// 화이트리스트면 그냥 통과
        if ((token == null || token.isBlank()) && isWhitelisted(request)) {
            authenticateAsGuest(request);
            chain.doFilter(request, response);
            return;
        }
```

yuyoungrhee replied:
@coderabbitai 답변해줘

yuyoungrhee committed:
[BE] test: 테스트 검증 추가 및 DisplayName 알맞게 변경 #626

yuyoungrhee commented:
@coderabbitai 치명적인 오류 및 가독성&일관성&리팩터링 필요한부분 등 리뷰

gabean13 commented on backend/app/src/main/java/com/onair/hearit/app/recommendation/application/CategoryRecommender.java:
```diff
@@ -0,0 +1,93 @@
+        recommendations.addAll(getRandomRecommendations(randomCount, recommendations));
+        return recommendations;
+    }
+
+    private Category getItTrendCategory() {
+        return categoryRepository.findByName(IT_TREND_CATEGORY_NAME)
+                .orElseThrow(() -> new NotFoundException("category", IT_TREND_CATEGORY_NAME));
+    }
```
findById로 하는 거 어떤가요? pk는 변할일이 없지만 it 트렌드 제목은 변할 가능성이 있을 것 같아요

yuyoungrhee replied:
prod랑 dev랑 로컬이랑 모두 IT 트렌드의 pk가 다를 수 있다고 생각했음요
지금은 물론 prod걸 dev로 한 번 백업한 경험이 있어서 둘 환경은 똑같지만 로컬 테스트 시 더 깨지기 쉬운 로직이 될 듯요.. 저도 이부분 고민했는데 어케해야 더 안전성있을지 모르겠음요 ㅜㅜ

gabean13 commented on backend/app/src/main/java/com/onair/hearit/app/recommendation/presentation/dto/RecommendationRequest.java:
```diff
@@ -0,0 +1,26 @@
+package com.onair.hearit.app.recommendation.presentation.dto;
+
+public record RecommendationRequest(
```
기존 dto패키지가 아니라 presentation안에 dto로 분리하신 이유가 있나용?

yuyoungrhee replied:
기존 dto 패키지는presentation, application 계층 모두 공통으로 사용하는 패키지 입니다.
반면에 RecommendationRequest의 역할은 HTTP 요청 파라미터를 컨트롤러까지 전달하는 것입니다. application 계층까지 전달되지 않아요. 그래서 오직 presentation 계층에만 의존하기 때문에 분리했습니다. 
<img width="959" height="297" alt="image" src="https://github.com/user-attachments/assets/6cda89b7-3e3e-4ac6-967e-54feffaee502" />

바깥 dto: 여러 계층에 걸쳐 사용하는 느낌을 줌
presentation/dto: presentation 계층에만 사용하는 느낌을 줌


근데 너무 일관성이 떨어지는 거 같으면 합쳐도 상관없긴 함요 어떻게 생각하세요?

gabean13 commented on backend/app/src/main/java/com/onair/hearit/app/recommendation/presentation/dto/RecommendationRequest.java:
```diff
@@ -0,0 +1,26 @@
+package com.onair.hearit.app.recommendation.presentation.dto;
+
+public record RecommendationRequest(
+        Integer categorySize,
+        Integer hearitSize
```
얘 int로 하는거 어떤가요? parmeter로 넘길 때 int로 넘기면 unboxing해야할텐데 그러면 그냥 첨부터 int로 설정해도 되지않을까욧

yuyoungrhee replied:
int로 하게 되면 null을 받을 수 없어서 null일 때 BindException이 발생합니다.

`@RequestParam`의 defaultValue를 활용하면 null일 시 int로 할당해주어 문제 없는데, 파라미터를 dto로 받는 경우 이 기능을 직접 구현해줘야해서 Integer 타입이 필요했습니다!

저는 사이즈가 Min/Max를 넘었을 때 예외가 아닌 defaultValue로 되돌리는 경험을 원했어서 단순 `@RequestParam`의 기능으로는 부족하더라고요.

gabean13 reviewed (commented):
멍구 고생했어욧

Byesol commented on backend/app/src/test/java/com/onair/hearit/app/recommendation/presentation/RecommendationIntegrationTest.java:
```diff
@@ -0,0 +1,138 @@
+                .getList(".", RecommendationByCategoryResponse.class);
+
+        // thena
+        SoftAssertions.assertSoftly(softly -> {
```
위에 //thena 오타요, 그리고 여기선 SoftAssertions 쓴 이유가 뭐에요?

yuyoungrhee replied:
get이 많아지면서 어떤 걸 검증하고 싶은지 가독성이 안 좋아져서 
`RecommendationByCategoryResponse exampleSingleResponse = responses.get(0);` << 얘를 중간에 변수로 정의해주고 싶었어요

`responses.get(0).hearits().get(0).hearitId()` -> `exampleSingleResponse.hearits().get(0).hearitId()`

SoftAssertions에서는 구문 사이에 이렇게 로직이 들어갈 수 있는데 assertAll은 못해서 SoftAssertsions 활용했습니다.

Byesol commented on backend/app/src/main/java/com/onair/hearit/app/recommendation/application/RecommendationService.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    private RecommendationByCategoryResponse toRecommendationByCategoryResponse(Category category, int hearitSize) {
+        List<Hearit> hearits = hearitRepository.findByCategory(category.getId(), hearitSize);
```
원래 기존에 있던 로직이긴 한데 지금 생각해보니 이것도  카테고리 개수만큼 Hearit 조회 쿼리를  실행하니 n+1 발생해서
hearitRepository에  categoryid들 한꺼번에 다 줘서 모든 hearit 한 번에 가져오는 게 더 낫긴 할듯요.

근데 어차피 홈화면 api 들은 계속 수정예정이니 놔둬도 괜찮아용

yuyoungrhee replied:
네엡 그러네요 이슈파서 해결해야겠습니다

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/app/recommendation/presentation/RecommendationController.java:
```diff
@@ -0,0 +1,33 @@
+            @AuthenticationPrincipal RequestUser requestUser,
+            RecommendationRequest request) {
+        List<RecommendationByCategoryResponse> responses =
+                recommendationService.getCategoryRecommendations(
```
꼭 한줄 띄울 필요 없을 것 같아요! (just 의견)

yuyoungrhee replied:
안 띄면 파라미터를 감싼 dto인 걸 헷갈릴 거 같아서,
ModelAttribute 어노테이션을 붙이면서 좀 더 명확하게 해보려고 했어욥
[[BE] refactor: 파라미터용 dto에 ModelAttribute 어노테이션 명시](https://github.com/woowacourse-teams/2025-hEARit/pull/636/commits/e1dd9ba11d088e36f44bc96620dc876504129571) https://github.com/woowacourse-teams/2025-hEARit/issues/626

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/app/recommendation/application/RecommendationService.java:
```diff
@@ -0,0 +1,32 @@
+@RequiredArgsConstructor
+public class RecommendationService {
+
+    private final RecommendCategoryService recommendCategoryService;
```
추천 로직 캡슐화 좋은데요?!
추천 서비스를 도와주는 클래스라면 `Service`보다는 Provider나 Picker? 이런 느낌은 어떤가요?
Service가 Service를 의존한다는게 지금 컨벤션에서는 어색하게 느껴집니다!

yuyoungrhee replied:
수정했습니다!
[[BE] refactor: 카테고리 추천 로직 클래스명 변경](https://github.com/woowacourse-teams/2025-hEARit/pull/636/commits/eac64f9a8baad7519fec0f3b76162b42572ca1b9) https://github.com/woowacourse-teams/2025-hEARit/issues/626

Byesol commented on backend/app/src/main/java/com/onair/hearit/app/recommendation/application/RecommendationService.java:
```diff
@@ -0,0 +1,32 @@
+@RequiredArgsConstructor
+public class RecommendationService {
+
+    private final RecommendCategoryService recommendCategoryService;
```
현재 RecommendCategoryService 와 RecommendationService 를 분리한 건 책임 분리 측면에서 엄청 좋다고 느껴지는데

그런데 이렇게 서비스의 책임을 분리하고 서비스가 서비스를 의존하는 형태로 둘 거라면

일단 저는 지금의 구조가 좋다고 생각들긴 하지만,  다른 관점으로 보면 이렇게 서비스 분리를 했다면 카테고리별 히어릿을 조회하는 책임도 이 서비스에서 분리해야하지 않나? 라는 생각도 들 수 있겠다 싶어서

 이번 기회에 기준을 정하면 좋을 것 같아요,

단순히 카테고리별 히어릿 조회를 위해 서비스를 분리한다는 건 모든 crud를 일일이 도메인 서비스로 분리하다는 건데 그렇게 하면 도메인 서비스가 bypass 역할만 하게 되어 큰 의미가 없다고 생각해서요.
 
그렇게 'crud만을 위한 도메인 서비스 분리는 좋지 않다.' 라는 기준은 세웠지만 '언제 서비스를 분리해야하는가?, 서비스끼리 계층은 분리해야한느가? ' 에 대한 기준은 아직 모호한 상태에요.
 
 따라서 제 기준은 모호하지만  ' 복잡성, 재사용성 의 가능성이 있을 때 서비스의 로직을 분리하고 domain 서비스랑 application 서비스 계층을 나눈다.' 인데 

 서비스 분리 및 의존에 대한 기준이 궁금해요.

yuyoungrhee replied:
아 제가 Service가 Service에 의존하는 구조를 만드려고 한 건 아니고 패키지를 옮기면서 좀 혼란이 발생했네요

RecommendCategoryService는 Component에 가깝습니다. 수정해볼게요
Service -> Service 의존 얘기는 나중에 각 패키지의 service끼리 의존해야할 때 다시 얘기해봐도 충분한 거 같아요
우선 지금 이슈와는 조금 다른 결인 것 같습니다. 지금 구조는 마치 explore 로직에서 상위Service가 있고 해당 책임을 여러 Component 클래스들로 나눈 거랑 오히려 더 비슷합니다.

[[BE] refactor: 카테고리 추천 로직 클래스명 변경](https://github.com/woowacourse-teams/2025-hEARit/pull/636/commits/eac64f9a8baad7519fec0f3b76162b42572ca1b9) https://github.com/woowacourse-teams/2025-hEARit/issues/626

Byesol reviewed (commented):
멍구 고생하셨어용!!

JO-eusan reviewed (changes requested):
멍고 연휴에 고생하셨어요!
간단한 리뷰 남겨두었습니다~

yuyoungrhee committed:
[BE] test: 테스트 오타 수정 #626

yuyoungrhee committed:
[BE] refactor: 파라미터용 dto에 ModelAttribute 어노테이션 명시 #626

yuyoungrhee committed:
[BE] erge branch 'develop-be' into feat/#626-add-it-trend-and-random-categories-to-recommended-hearit #626

# Conflicts:
#	backend/app/src/main/java/com/onair/hearit/app/auth/config/ApiSecurityConfig.java
#	backend/app/src/main/java/com/onair/hearit/app/hearit/application/HearitService.java
#	backend/app/src/main/java/com/onair/hearit/app/hearit/presentation/HearitController.java
#	backend/app/src/test/java/com/onair/hearit/app/hearit/application/HearitServiceTest.java
#	backend/app/src/test/java/com/onair/hearit/app/hearit/presentation/HearitControllerTest.java
#	backend/app/src/test/java/com/onair/hearit/app/hearit/presentation/HearitIntegrationTest.java

yuyoungrhee committed:
[BE] refactor: 카테고리 추천 로직 클래스명 변경 #626

- RecommendCategoryService -> CategoryRecommender

gabean13 reviewed (approved):
고생햇으예

JO-eusan reviewed (approved):
커멘트 확인했습니다. Approve~
